src/emulator/cpu/src/unicorn_cpu.cpp

```diff
@@ -0,0 +1,388 @@
+    assert(intno == 2);
+
+    CPUState &state = *static_cast<CPUState *>(user_data);
+
+    uint32_t pc = 0;
+    uc_err err = uc_reg_read(uc, UC_ARM_REG_PC, &pc);
+    assert(err == UC_ERR_OK);
+
+    switch (intno) {
+    case INT_SVC:
+        if (thumb_mode(uc)) {
+            const Address svc_address = pc - 2;
+            uint16_t svc_instruction = 0;
+            uc_err err = uc_mem_read(uc, svc_address, &svc_instruction, sizeof(svc_instruction));
+            assert(err == UC_ERR_OK);
+            const uint8_t imm = svc_instruction & 0xff;
+            state.call_svc(state, imm, pc);
+        } else {
+            const Address svc_address = pc - 4;
+            uint32_t svc_instruction = 0;
+            uc_err err = uc_mem_read(uc, svc_address, &svc_instruction, sizeof(svc_instruction));
+            assert(err == UC_ERR_OK);
+            const uint32_t imm = svc_instruction & 0xffffff;
+            state.call_svc(state, imm, pc);
+        }
+        break;
+    case INT_BKPT:
+        stop(state);
+        state.did_break = true;
+        break;
+    default: break;
+    }
+}
+
+static void enable_vfp_fpu(uc_engine *uc) {
+    uint64_t c1_c0_2 = 0;
+    uc_err err = uc_reg_read(uc, UC_ARM_REG_C1_C0_2, &c1_c0_2);
+    assert(err == UC_ERR_OK);
+
+    c1_c0_2 |= (0xf << 20);
+
+    err = uc_reg_write(uc, UC_ARM_REG_C1_C0_2, &c1_c0_2);
+    assert(err == UC_ERR_OK);
+
+    const uint64_t fpexc = 0xf0000000;
+
+    err = uc_reg_write(uc, UC_ARM_REG_FPEXC, &fpexc);
+    assert(err == UC_ERR_OK);
+}
+
+UnicornCPU::UnicornCPU(CPUState *state, Address pc, Address sp, bool log_code) {
+    uc_engine *temp_uc = nullptr;
+    uc_err err = uc_open(UC_ARCH_ARM, UC_MODE_ARM, &temp_uc);
+    assert(err == UC_ERR_OK);
+
+    uc = UnicornPtr(temp_uc, uc_free);
+
+    temp_uc = nullptr;
+    uc_hook hh = 0;
+
+    err = uc_hook_add(uc.get(), &hh, UC_HOOK_INTR, reinterpret_cast<void *>(&intr_hook), state, 1, 0);
+    assert(err == UC_ERR_OK);
+
+    if ((log_code && LOG_CODE) || LOG_CODE_ALL) {
+        uc_hook hh = 0;
+        const uc_err err = uc_hook_add(uc.get(), &hh, UC_HOOK_CODE, reinterpret_cast<void *>(&code_hook), &state, 1, 0);
+
+        assert(err == UC_ERR_OK);
+    }
+
+    if (LOG_MEM_ACCESS) {
+        uc_hook hh = 0;
+        uc_err err = uc_hook_add(uc.get(), &hh, UC_HOOK_MEM_READ, reinterpret_cast<void *>(&read_hook), &state, 1, 0);
+        assert(err == UC_ERR_OK);
+
+        err = uc_hook_add(uc.get(), &hh, UC_HOOK_MEM_WRITE, reinterpret_cast<void *>(&write_hook), &state, 1, 0);
+        assert(err == UC_ERR_OK);
+    }
+
+    err = uc_reg_write(uc.get(), UC_ARM_REG_SP, &sp);
+    assert(err == UC_ERR_OK);
+
+    err = uc_mem_map_ptr(uc.get(), 0, GB(4), UC_PROT_ALL, &(state->mem->memory[0]));
+    assert(err == UC_ERR_OK);
+
+    err = uc_reg_write(uc.get(), UC_ARM_REG_PC, &pc);
+
+    assert(err == UC_ERR_OK);
+
+    err = uc_reg_write(uc.get(), UC_ARM_REG_LR, &pc);
+
+    assert(err == UC_ERR_OK);
```

Comment:
man we really need a `UC_CHECK` macro or something instead of all these asserts (not saying you have to do it on this PR)